### PR TITLE
Versioning for changes to Ladybug Tools Toolkit

### LINKED ISSUE #240: Versioning for LBT TK
#### Broken rules:

A number of changes to objects in https://github.com/BHoM/LadybugTools_Toolkit/pull/128 break versioning and need converters.

#### Suggestions to restore compliance:

Update the converter to fix versioning for above PR

### PR DESCRIPTION
**Pull request by Tom-Kingstone**
### NOTE: Depends on 

https://github.com/BHoM/LadybugTools_Toolkit/pull/128

   
### Issues addressed by this PR

Closes #240

### CODE REVIEW

File: BHoMUpgrader70/Converter.cs [hunk 1]
```diff
@@ -39,6 +39,158 @@ public class Converter : Base.Converter
...
+                }
+                if ((string)oldVersion["_t"] == "BH.oM.LadybugTools.ILadybugToolsMaterial")
+                {
+                    newVersion["_t"] = "BH.oM.LadybugTools.IEnergyMaterialOpaque";
+                }
+            }
+
+            if (newVersion.ContainsKey("Source"))
+            {
+                newVersion["Source"] = null;
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:95]:
```suggestion
```

File: BHoMUpgrader70/Converter.cs [hunk 2]
```diff
@@ -39,6 +39,158 @@ public class Converter : Base.Converter
...
+            if (newVersion.TryGetValue("WindPorosity", out object wind))
+            {
+                numbers = Enumerable.Repeat((double)wind, 8760).ToList();
+                newVersion["WindPorosity"] = numbers;
+            }
+            if (newVersion.TryGetValue("RadiationPorosity", out object radiation))
+            {
+                numbers = Enumerable.Repeat((double)radiation, 8760).ToList();
+                newVersion["RadiationPorosity"] = numbers;
+            }
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:68]:
```suggestion
            if (newVersion.TryGetValue("WindPorosity", out object wind))
            {
                numbers = Enumerable.Repeat((double)wind, 8760).ToList();
                newVersion["WindPorosity"] = numbers;
            }
            
            if (newVersion.TryGetValue("RadiationPorosity", out object radiation))
            {
                numbers = Enumerable.Repeat((double)radiation, 8760).ToList();
                newVersion["RadiationPorosity"] = numbers;
            }
```

File: BHoMUpgrader70/Converter.cs [hunk 3]
```diff
@@ -39,6 +39,158 @@ public class Converter : Base.Converter
...
+            if (newVersion.ContainsKey("Source"))
+            {
+                newVersion["Source"] = null;
+                newVersion.Remove("Source");
+            }
+
+            return newVersion;
+        }
+
+        private static Dictionary<string, object> HourlyContinuousCollection(Dictionary<string, object> CustomData)
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:102]: Little c on `customData` variable please

File: BHoMUpgrader70/Converter.cs [hunk 4]
```diff
@@ -39,6 +39,158 @@ public class Converter : Base.Converter
...
+        {
+            Dictionary<string, object> newVersion = new Dictionary<string, object>();
+
+            newVersion.Add("_t", "BH.oM.LadybugTools.HourlyContinuousCollection");
+            newVersion.Add("Type", CustomData["Type"]);
+            newVersion.Add("Header", CustomData["Header"]);
+            newVersion.Add("Values", CustomData["Values"]);
+
+            Dictionary<string, object> Header = newVersion["Header"] as Dictionary<string, object>;
+            Header.Add("_t", "BH.oM.LadybugTools.Header");
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:112]:
```suggestion
            Dictionary<string, object> header = newVersion["Header"] as Dictionary<string, object>;
           header.Add("_t", "BH.oM.LadybugTools.Header");
```

File: BHoMUpgrader70/Converter.cs [hunk 5]
```diff
@@ -39,6 +39,158 @@ public class Converter : Base.Converter
...
+
+            (Header["Datatype"] as Dictionary<string, object>).Add("_t", "BH.oM.LadybugTools.DataType");
+            (Header["AnalysisPeriod"] as Dictionary<string, object>).Add("_t", "BH.oM.LadybugTools.AnalysisPeriod");
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:115]:
```suggestion
            (header["Datatype"] as Dictionary<string, object>).Add("_t", "BH.oM.LadybugTools.DataType");
            (header["AnalysisPeriod"] as Dictionary<string, object>).Add("_t", "BH.oM.LadybugTools.AnalysisPeriod");
```

File: BHoMUpgrader70/Converter.cs [hunk 6]
```diff
@@ -39,6 +39,158 @@ public class Converter : Base.Converter
...
+
+            newVersion["Header"] = Header;
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:117]:
```suggestion
            newVersion["Header"] = header;
```

File: BHoMUpgrader70/Converter.cs [hunk 7]
```diff
@@ -39,6 +39,158 @@ public class Converter : Base.Converter
...
+            return newVersion;
+        }
+
+        //So much stuff to version!
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:121]:
```suggestion
```

😉

File: BHoMUpgrader70/Converter.cs [hunk 8]
```diff
@@ -39,6 +39,158 @@ public class Converter : Base.Converter
...
+        private static Dictionary<string, object> UpgradeExternalComfort(Dictionary<string, object> oldVersion)
+        {
+            Dictionary<string, object> newVersion = new Dictionary<string, object>(oldVersion);
+            Dictionary<string, object> CustomData;
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:125]: Little `c` please.

File: BHoMUpgrader70/Converter.cs [hunk 9]
```diff
@@ -39,6 +39,158 @@ public class Converter : Base.Converter
...
+                }
+            }
+
+            return newVersion;
+        }
+
+        private static Dictionary<string, object> UpgradeSimulationResult(Dictionary<string, object> oldVersion)
+        {
+            Dictionary<string, object> newVersion = new Dictionary<string, object>(oldVersion);
+            Dictionary<string, object> CustomData;
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:154]: Little `c`

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]
Can you also put the method separators between each method too please 😄

**Comment by Tom-Kingstone** [comment]
@BHoMBot check versioning

File: BHoMUpgrader70/Converter.cs [hunk 10]
```diff
@@ -39,6 +39,165 @@ public class Converter : Base.Converter
...
+        /***************************************************/
+
+        private static Dictionary<string, object> HourlyContinuousCollection(Dictionary<string, object> customData)
+        {
+            Dictionary<string, object> newVersion = new Dictionary<string, object>();
+
+            newVersion.Add("_t", "BH.oM.LadybugTools.HourlyContinuousCollection");
+            newVersion.Add("Type", customData["Type"]);
+            newVersion.Add("Header", customData["Header"]);
+            newVersion.Add("Values", customData["Values"]);
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:113]:
```suggestion
            newVersion.Add("Values", customData["Values"] as List<double>);
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Comment by FraserGreenroyd** [comment]
@BHoMBot check versioning

File: BHoMUpgrader70/Converter.cs [hunk 11]
```diff
@@ -39,6 +39,165 @@ public class Converter : Base.Converter
...
+        }
+
+        /***************************************************/
+
+        private static Dictionary<string, object> HourlyContinuousCollection(Dictionary<string, object> customData)
+        {
+            Dictionary<string, object> newVersion = new Dictionary<string, object>();
+
+            newVersion.Add("_t", "BH.oM.LadybugTools.HourlyContinuousCollection");
+            newVersion.Add("Type", customData["Type"]);
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:111]:
```suggestion
            if(customData.ContainsKey("Type"))
                newVersion.Add("Type", customData["Type"]);
```

File: BHoMUpgrader70/Converter.cs [hunk 12]
```diff
@@ -39,6 +39,165 @@ public class Converter : Base.Converter
...
+            newVersion.Add("Header", customData["Header"]);
+            newVersion.Add("Values", customData["Values"] as List<double>);
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:113]:
```suggestion
            if(customData.ContainsKey("Header"))
                newVersion.Add("Header", customData["Header"]);
           if(customData.ContainsKey("Values"))
                newVersion.Add("Values", customData["Values"] as List<double>);
           
```

File: BHoMUpgrader70/Converter.cs [hunk 13]
```diff
@@ -39,6 +39,165 @@ public class Converter : Base.Converter
...
+
+            Dictionary<string, object> header = newVersion["Header"] as Dictionary<string, object>;
+            header.Add("_t", "BH.oM.LadybugTools.Header");
+
+            (header["Datatype"] as Dictionary<string, object>).Add("_t", "BH.oM.LadybugTools.DataType");
+            (header["AnalysisPeriod"] as Dictionary<string, object>).Add("_t", "BH.oM.LadybugTools.AnalysisPeriod");
+
+            newVersion["Header"] = header;
```
Comment (FraserGreenroyd) [inline BHoMUpgrader70/Converter.cs:121]:
```suggestion
            if(newVersion.ContainsKey("Header"))
            {
                Dictionary<string, object> header = newVersion["Header"] as Dictionary<string, object>;
                header.Add("_t", "BH.oM.LadybugTools.Header");

                (header["Datatype"] as Dictionary<string, object>).Add("_t", "BH.oM.LadybugTools.DataType");
                (header["AnalysisPeriod"] as Dictionary<string, object>).Add("_t", "BH.oM.LadybugTools.AnalysisPeriod");

                newVersion["Header"] = header;
            }
```

**Review by FraserGreenroyd — CHANGES_REQUESTED** [review]

**Comment by Tom-Kingstone** [comment]
@BHoMBot check versioning

**Comment by FraserGreenroyd** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check core

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: installer

**Comment by FraserGreenroyd** [comment]
@BHoMBot this is a DevOps instruction. I am requesting neutral checks on: ready-to-merge

**Review by FraserGreenroyd — APPROVED** [review]
Good work @Tom-Kingstone 😄

**Pull request merged**